### PR TITLE
[TELEGRAM]: core notification

### PR DESCRIPTION
**Pull request by keda-vlad**
added daily notification for overdue
added notification for new created rental

### CODE REVIEW

File: src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java [hunk 1]
```diff
@@ -1,17 +1,78 @@
...
+    private final RentalRepository rentalRepository;
 
     @Override
     public void sendNotification(String string) {
         bot.sendMessage(string);
     }
+
+    @Scheduled(cron = CRON)
+    private void sendDailyStatistic() {
+        LocalDate date = LocalDate.now().plusDays(1);
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java:39]: Magic number 🪄

**Review by Denis-Balako — APPROVED** [review]
Great!

File: src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java (hunk 1 above)
Comment (Someboty) [reply src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java:39]: And do you need that variable separately?

File: src/main/java/com/team01/carsharingapp/repository/RentalRepository.java [hunk 2]
```diff
@@ -43,4 +44,13 @@ public interface RentalRepository extends JpaRepository<Rental, Long> {
            OR (:isActive = false AND r.actualReturnDate IS NOT NULL))
             """)
     List<Rental> findAllByUserIdAndStatus(Long userId, boolean isActive, Pageable pageable);
+
+    @Query(""" 
+           FROM Rental r
+           LEFT JOIN FETCH r.car c
+           LEFT JOIN FETCH r.user u
+           WHERE r.actualReturnDate IS NULL
+           AND r.rentalDate > :date
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/repository/RentalRepository.java:53]: Why do you need to compare rentalDate and date? Maybe it should be returnDate?

File: src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java [hunk 3]
```diff
@@ -1,17 +1,78 @@
...
+        sendNotification(message);
+    }
+
+    @EventListener
+    private void sendCreatedRental(CreateRentalEvent createRentalEvent) {
+        sendNotification(buildCreatedRentalMessage(createRentalEvent.getRental()));
+    }
+
+    private String buildCreatedRentalMessage(Rental rental) {
+        String user = rental.getUser().getEmail();
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java:51]: So it's user or email? Bad naming.

File: src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java [hunk 4]
```diff
@@ -1,17 +1,78 @@
...
+        String date = rental.getReturnDate().toString();
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java:52]: Maybe returnDate, but it's optional

File: src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java [hunk 5]
```diff
@@ -1,17 +1,78 @@
...
+        String car = rental.getCar().getBrand() + " " + rental.getCar().getModel();
+        return String.format(CREATE_RENTAL_MESSAGE, user, date, car);
+    }
+
+    private String buildOverdueMessage(List<Rental> overdue) {
+        Function<Rental, String> buildMessage = r -> {
+            String user = r.getUser().getEmail();
+            String date = r.getReturnDate().toString();
+            String car = r.getCar().getBrand() + " " + r.getCar().getModel();
+            return String.format(OVERDUE_MESSAGE, user, date, car);
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java:62]: It's just your buildCreatedRentalMessage() method. Can you call it here?

File: src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java [hunk 6]
```diff
@@ -1,17 +1,78 @@
...
+        };
+        Comparator<Rental> comparator = (f, s) -> {
+            if (f.getReturnDate().isBefore(s.getRentalDate())) {
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/TelegramNotificationService.java:65]: How return date can be before rental date?

**Review by Someboty — CHANGES_REQUESTED** [review]
GJ! Please look at the suggestions.

**Review by ConfusedSatlan — APPROVED** [review]

**Pull request merged**